core/node/commitment_generator/src/lib.rs

```diff
@@ -21,34 +21,53 @@ use zksync_utils::h256_to_u256;
 
 use crate::{
     metrics::{CommitmentStage, METRICS},
-    utils::{bootloader_initial_content_commitment, events_queue_commitment},
+    utils::{CommitmentComputer, RealCommitmentComputer},
 };
 
 mod metrics;
+#[cfg(test)]
+mod tests;
 mod utils;
 pub mod validation_task;
 
 const SLEEP_INTERVAL: Duration = Duration::from_millis(100);
 
+/// Component responsible for generating commitments for L1 batches.
 #[derive(Debug)]
 pub struct CommitmentGenerator {
+    computer: Arc<dyn CommitmentComputer>,
     connection_pool: ConnectionPool<Core>,
     health_updater: HealthUpdater,
     commitment_mode: L1BatchCommitmentMode,
+    parallelism: NonZeroU32,
 }
 
 impl CommitmentGenerator {
+    /// Creates a commitment generator with the provided mode.
     pub fn new(
         connection_pool: ConnectionPool<Core>,
         commitment_mode: L1BatchCommitmentMode,
     ) -> Self {
         Self {
+            computer: Arc::new(RealCommitmentComputer),
             connection_pool,
             health_updater: ReactiveHealthCheck::new("commitment_generator").1,
             commitment_mode,
+            parallelism: Self::default_parallelism(),
         }
     }
 
+    fn default_parallelism() -> NonZeroU32 {
+        // Leave at least one core free to handle other blocking tasks. `unwrap()`s are safe by design.
+        let cpus = u32::try_from(num_cpus::get().saturating_sub(1).clamp(1, 16)).unwrap();
```

Comment:
`process_batch` spawns 2 blocking tasks: for `events_queue` and `bootloader_memory`. I'm wondering how that will work when parallelism is set to `num_cpus::get().saturating_sub(1)`. I think it should be fine, especially given that events task usually takes up to 1s, while bootloader task takes around 3s for v22 and 6s for v24; but still want to double-check there won't be issues because of that

Reply:
What kind of issues do you have in mind? Since parallelism only controls the max number of L1 batches processed at the same time (not, say, the number of blocking threads for Tokio), it doesn't look possible to deadlock generation if the parallelism factor is too large.